python/ray/autoscaler/_private/spark/spark_job_server.py

```diff
@@ -0,0 +1,176 @@
+                    _logger.warning(
+                        f"Spark job {spark_job_group_id} hosting Ray worker node exit."
+                    )
+
+            threading.Thread(
+                target=inheritable_thread_target(start_ray_worker_thread_fn),
+                args=(),
+                daemon=True,
+            ).start()
+
+            self.server.task_status_dict[spark_job_group_id] = "pending"
+            return {}
+
+        elif path_parts[0] == "terminate_node":
+            assert len(path_parts) == 1, f"Illegal request path: {path}"
+            self.server.spark.sparkContext.cancelJobGroup(spark_job_group_id)
+            if spark_job_group_id in self.server.task_status_dict:
+                self.server.task_status_dict.pop(spark_job_group_id)
+            return {}
+
+        elif path_parts[0] == "notify_task_launched":
+            if spark_job_group_id in self.server.task_status_dict:
+                # Note that if `spark_job_group_id` not in task_status_dict,
+                # the task has been terminated
+                self.server.task_status_dict[spark_job_group_id] = "running"
+                _logger.info(f"Spark task in {spark_job_group_id} has started.")
+            return {}
+
+        elif path_parts[0] == "query_task_status":
+            if spark_job_group_id in self.server.task_status_dict:
+                return {"status": self.server.task_status_dict[spark_job_group_id]}
+            else:
+                return {"status": "terminated"}
+
+        else:
+            raise ValueError(f"Illegal request path: {path}")
+
+    def do_POST(self):
+        """Reads post request body"""
+        self._set_headers()
+        content_len = int(self.headers["content-length"])
+        content_type = self.headers["content-type"]
+        assert content_type == "application/json"
+        path = self.path
+        post_body = self.rfile.read(content_len).decode("utf-8")
+        post_body_json = json.loads(post_body)
+        with self._handler_lock:
+            response_body_json = self.handle_POST(path, post_body_json)
+        response_body = json.dumps(response_body_json)
+        self.wfile.write(response_body.encode("utf-8"))
+
+    def log_request(self, code="-", size="-"):
+        # Make logs less verbose.
+        pass
+
+
+class SparkJobServer(ThreadingHTTPServer):
+    """
+    The http server that is used to launch spark jobs for holding ray worker nodes.
+    Because we have to use "spark session" instance to create spark job,
+    and "spark session" instance only exists inside spark application driver
+    process, and ray on spark cluster head node process is a subprocess of
+    spark application driver process,
+    but Ray autoscaler is executed as sub-process of ray head node process,
+    so we create the spark job server inside spark application driver
+    process, and when ray autoscaler requests to create spark jobs, it sends
+    http request to the spark job server.
```

Comment:
Can you split into multiple sentences. Currently it's a bit hard to understand. Also this doesn't explain why we have to use the "spark session" instance in the spark application driver process. One way is to reference to the REP.